homeassistant/components/canary/camera.py

```diff
@@ -125,17 +130,40 @@ async def async_camera_image(
         self, width: int | None = None, height: int | None = None
     ) -> bytes | None:
         """Return a still image response from the camera."""
-        await self.hass.async_add_executor_job(self.renew_live_stream_session)
-        live_stream_url = await self.hass.async_add_executor_job(
-            getattr, self._live_stream_session, "live_stream_url"
-        )
-        return await ffmpeg.async_get_image(
-            self.hass,
-            live_stream_url,
-            extra_cmd=self._ffmpeg_arguments,
-            width=width,
-            height=height,
-        )
+        await self._check_for_image_expiration()
+
+        if self._image is None:
+            _LOGGER.debug("Grabbing a live view image from %s", self.name)
+            await self.hass.async_add_executor_job(self.renew_live_stream_session)
+
+            if (live_stream_session := self._live_stream_session) is None:
+                return None
+
+            if not (live_stream_url := live_stream_session.live_stream_url):
+                return None
+
+            image = await ffmpeg.async_get_image(
+                self.hass,
+                live_stream_url,
+                extra_cmd=self._ffmpeg_arguments,
+                width=width,
+                height=height,
+            )
+
+            if image:
+                self._image = image
+                _LOGGER.debug("Grabbed a live view image from %s", self.name)
+            await self.hass.async_add_executor_job(live_stream_session.stop_session)
+            _LOGGER.debug("Stopped live session from %s", self.name)
+
+        return self._image
+
+    async def _check_for_image_expiration(self) -> None:
+        utcnow = dt_util.utcnow()
+        if self._expires_at <= utcnow:
+            _LOGGER.debug("Expiring cached image from %s", self.name)
+            self._image = None
+            self._expires_at = FORCE_CAMERA_REFRESH_INTERVAL + utcnow
```

Comment:
I realize its somewhat arbitrary when  you set the expiration time, but logically to me it makes the most sense to set `_expires_at` in the same block you set `_image`.

I would suggest that it might be easiest to just inline the `expires_at` check  on the same line as the `if self._image is None:` check then also update the expiration time after grabbing the image.  In general, I do like separation of responsibilities, but i think here its easier to reason about all in one function.

Reply:
done; now has the added benefit of not clearing the image prior to getting a new one. This way if there is an issue getting the new image it will not display a broken image.